### PR TITLE
[MXNET-1258]fix unittest for ROIAlign Operator

### LINKED ISSUE #11064: Flaky test: test_operator.test_op_roi_align
```
======================================================================

FAIL: test_operator.test_op_roi_align

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/local/lib/python3.5/dist-packages/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/work/mxnet/tests/python/unittest/common.py", line 157, in test_new

    orig_test(*args, **kwargs)

  File "/work/mxnet/tests/python/unittest/test_operator.py", line 6170, in test_op_roi_align

    test_roi_align_value()

  File "/work/mxnet/tests/python/unittest/test_operator.py", line 6149, in test_roi_align_value

    assert np.allclose(data.grad.asnumpy(), dx, atol = 1e-6), np.abs(data.grad.asnumpy() - dx).max()

AssertionError: 1.3150275e-06

-------------------- >> begin captured logging << --------------------

common: INFO: Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1619190489 to reproduce.

--------------------- >> end captured logging << ---------------------

```
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11058/1/pipeline

### PR DESCRIPTION
**Pull request by wkcn**
## Description ##
Fix ROIAlign unittest #11064 
It passes the test whose `MXNET_TEST_SEED=35650200`. 

The old unittest for ROIAlign computes in `float64`, so there is some float precision problem.
Fixes #11064
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] use `float32` in the unittest for ROIAlign OP
- [x] set `atol` to `1e-5` in the unittest for ROIAlign OP

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by wkcn** [comment]
It seems there is a problem in Windows-CPU CI.
`WindowsError: [Error 5] Access is denied: 'c:\\Anaconda3\\envs\\py2\\Lib\\site-packages\\six-1.12.0.dist-info\\RECORD.pip'
`

**Comment by roywei** [comment]
@mxnet-label-bot add[CI, Flaky, pr-awaiting-review]

**Comment by anirudhacharya** [comment]
can you run the test >500 times and paste the results here to ensure that the test is no longer flaky.

**Comment by wkcn** [comment]
@anirudhacharya Thank you. I will test it.

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -6845,30 +6845,32 @@ def test_context_num_gpus():
...
+
     def bilinear_interpolate(bottom, height, width, y, x):
         if y < -1.0 or y > height or x < -1.0 or x > width:
-            return 0.0, []
-        x = max(0.0, x)
-        y = max(0.0, y)
+            return T(0.0), []
+        x = T(max(0.0, x))
+        y = T(max(0.0, y))
         x_low = int(x)
```
[the remark this answers is not in the document]
Comment (wkcn) [reply tests/python/unittest/test_operator.py]: The type of x_low is `int` in [the C implementation](https://github.com/apache/incubator-mxnet/blob/master/src/operator/contrib/roi_align.cu).

In Python, the default type of float is `float64`, however the float type in the C implementation is `float32`.
If the types are not consistent, there will be calculation error.

So I use `T` to typecast some variables to `np.float32` to keep the type consistency between C and Python unittest.

File: tests/python/unittest/test_operator.py [hunk 2]
```diff
@@ -6891,54 +6894,70 @@ def bilinear_interpolate(bottom, height, width, y, x):
...
                                 # compute grad
                                 for qy, qx, qw in g:
-                                    dx[batch_ind, c, qy, qx] += dy[r, c, ph, pw] * qw * 1.0 / count
+                                    assert qw.dtype == T
+                                    dx[batch_ind, c, qy, qx] += dy[r,
+                                                                   c, ph, pw] * qw / count
 
-                        out[r, c, ph, pw] = val * 1.0 / count
+                        out[r, c, ph, pw] = val / count
+        assert out.dtype == T, out.dtype
```
[the remark this answers is not in the document]
Comment (wkcn) [reply tests/python/unittest/test_operator.py]: Thanks! I will add more detail.

File: tests/python/unittest/test_operator.py [hunk 3]
```diff
@@ -6845,30 +6845,32 @@ def test_context_num_gpus():
 
 @with_seed()
 def test_op_roi_align():
-    # Adapted from https://github.com/wkcn/MobulaOP/blob/master/tests/test_roi_align_op.py
+    # Adapted from https://github.com/wkcn/MobulaOP/blob/master/tests/test_op/test_roi_align_op.py
```
[the remark this answers is not in the document]
Comment (wkcn) [reply tests/python/unittest/test_operator.py]: I have removed it. Thanks!

**Comment by Roshrini** [comment]
@anirudhacharya Can you take a look again?

**Comment by anirudhacharya** [comment]
LGTM

**Comment by anirudhacharya** [comment]
@wkcn please rebase and resolve conflicts

@mxnet-label-bot update [pr-awaiting-merge]

**Comment by wkcn** [comment]
@anirudhacharya In my own project, there is also a ROIAlignOP unittest, and I found that nvcc will change the order of operators because of compiler optimization, causing the problem of float precision. So please not merge the PR temporarily, and I will check it.

**Comment by stu1130** [comment]
@mxnet-label-bot update [pr-work-in-progress]

**Comment by wkcn** [comment]
@stu1130 Thanks!

**Comment by wkcn** [comment]
test fail in test activation
'''
====================================================================

FAIL: test_operator.test_activation

----------------------------------------------------------------------

Traceback (most recent call last):

  File "C:\Anaconda3\envs\py3\lib\site-packages\nose\case.py", line 197, in runTest

    self.test(*self.arg)

  File "C:\jenkins_slave\workspace\ut-python-gpu\tests\python\unittest\common.py", line 173, in test_new

    orig_test(*args, **kwargs)

  File "C:\jenkins_slave\workspace\ut-python-gpu\tests\python\unittest\test_operator.py", line 6820, in test_activation

    name, op[0], shape, op[3], op[4], rtol_fd, atol_fd, num_eps)

  File "C:\jenkins_slave\workspace\ut-python-gpu\tests\python\unittest\test_operator.py", line 6086, in finite_diff_unary_op

    check_grad(op_ex, [data_np])

  File "C:\jenkins_slave\workspace\ut-python-gpu\tests\python\unittest\test_operator.py", line 6079, in <lambda>

    atol=atol, dtype=dtype)

  File "C:\jenkins_slave\workspace\ut-python-gpu\windows_package\python\mxnet\test_utils.py", line 921, in check_numeric_gradient

    ("NUMERICAL_%s"%name, "BACKWARD_%s"%name))

  File "C:\jenkins_slave\workspace\ut-python-gpu\windows_package\python\mxnet\test_utils.py", line 495, in assert_almost_equal

    raise AssertionError(msg)

AssertionError: 

Items are not equal:

Error 113966.931334 exceeds tolerance rtol=0.000010, atol=0.000001.  Location of maximum error:(0, 4, 8, 3), a=0.113967, b=0.000000

 NUMERICAL_data: array([[[[ 0.        ,  0.68042432,  0.        , ...,  0.34353021,

           0.13880596,  0.94525056],

         [ 0.11517657,  0.18770058,  0.30324909, ...,  0.97787645,...

 BACKWARD_data: array([[[[ 0.        ,  0.68042432,  0.        , ...,  0.34353021,

           0.13880596,  0.94525056],

         [ 0.11517657,  0.18770058,  0.30324909, ...,  0.97787645,...

-------------------- >> begin captured logging << --------------------

common: INFO: Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1553516940 to reproduce.

--------------------- >> end captured logging << ---------------------
'''

**Comment by wkcn** [comment]
@anirudhacharya @stu1130 
Hi! I think the PR is finished.
Since `nvcc` will change the order of operator, I increase the tolerance to `1e-3`.

**Comment by anirudhacharya** [comment]
@sandeep-krishnamurthy have already approved this PR, above.

**Pull request merged**